File: 52-n-queens-ii/52-n-queens-ii.py
class Solution:
    def __init__(self):
        self.count = 0
    
    def isSafe(self, board, idx, val):
        board[idx] = val
        for i in range(len(board)):
            if (i != idx and board[i] == val):
                board[idx] = -1
                return False
        
        for i in range(len(board)-1):
            for j in range(i+1, len(board)):
                if (board[i] != -1 and board[j] != -1 and abs(board[i] - board[j]) == j - i):
                    board[idx] = -1
                    return False
        board[idx] = -1
        return True
    
    def solve(self, board, idx):
        # print(board)
        if (idx == len(board)):
            self.count += 1
            return True
        
        for i in range(len(board)):
            if self.isSafe(board, idx, i):
                board[idx] = i
                self.solve(board, idx + 1)
                board[idx] = -1
            
        return False
    
    
    def totalNQueens(self, n: int) -> int:
        board = [-1 for _ in range(n)]
        self.solve(board, 0)
        return self.count

File: 52-n-queens-ii/52-n-queens-ii.cpp
class Solution {
public:
    
    bool isSafe(vector<string> &s, int x, int y)
    {
        int i, j;
        for (i=0; i<y; i++)
            if (s[x][i] == 'Q')
                return false;
        
        for (i=x, j=y; j>=0 && i>=0; i--, j--)
            if (s[i][j] == 'Q')
                return false;
        
        for (i=x, j=y; j>=0 && i<s.size(); i++, j--)
            if (s[i][j] == 'Q')
                return false;
        
        return true;
    }
    
    bool solve(int &res, vector<string> s, int n, int y)
    {
        if (y >= n)
        {
            res++;
            return true;
        }
        
        for (int x=0; x<n; x++)
        {
            if (isSafe(s, x, y))
            {
                s[x][y] = 'Q';
                if (!solve(res, s, n, y + 1))
                {
                    s[x][y] = '.';
                }
            }
        }
        return false;
    }
    
    int totalNQueens(int n) {
        int res=0;
        vector<string> s(n, string(n, '.'));
        solve(res, s, n, 0);
        return res;
    }
};
